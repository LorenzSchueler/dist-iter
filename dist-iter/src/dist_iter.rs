use mpi::{
    topology::SimpleCommunicator,
    traits::{Communicator, Source},
};

use crate::task::Task;

pub struct DistIter<'w, I>
where
    I: Iterator,
    I::Item: Task,
{
    inner: I,
    init: bool,
    send_count: usize,
    recv_count: usize,
    world: &'w SimpleCommunicator,
}

impl<'w, I> Iterator for DistIter<'w, I>
where
    I: Iterator,
    I::Item: Task,
{
    type Item = <I::Item as Task>::OUT;

    fn next(&mut self) -> Option<Self::Item> {
        if !self.init {
            for dest in 1..self.world.size() {
                if let Some(task) = self.inner.next() {
                    task.send(self.world.process_at_rank(dest));
                    self.send_count += 1;
                }
            }
        }
        if self.recv_count < self.send_count {
            let tag = <I::Item as Task>::TAG;
            let (data, status) = self.world.any_process().receive_with_tag(tag);
            self.recv_count += 1;

            if let Some(task) = self.inner.next() {
                task.send(self.world.process_at_rank(status.source_rank()));
                self.send_count += 1;
            }
            Some(data)
        } else {
            None
        }
    }
}

pub trait MyIterExt {
    fn into_dist_iter(self, world: &SimpleCommunicator) -> DistIter<Self>
    where
        Self: Iterator + Sized,
        Self::Item: Task;
}

impl<I> MyIterExt for I
where
    I: Iterator,
    I::Item: Task,
{
    fn into_dist_iter(self, world: &SimpleCommunicator) -> DistIter<Self> {
        DistIter {
            inner: self,
            init: false,
            send_count: 0,
            recv_count: 0,
            world,
        }
    }
}
