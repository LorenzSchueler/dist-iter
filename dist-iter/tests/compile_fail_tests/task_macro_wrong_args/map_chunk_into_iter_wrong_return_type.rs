use dist_iter::{map_chunk_task, DistIterator};

fn main() {
    [1, 2, 3, 4, 5].into_iter().dist_map_chunk(map_chunk_task!(
        INPUT_CHUNK_SIZE = 2,
        OUTPUT_CHUNK_SIZE = 2,
        |buf: impl Iterator<Item = i32>| -> impl IntoIterator<Item = u32> { buf.map(|x| x + 1) }
    ));
    [1, 2, 3, 4, 5].into_iter().dist_map_chunk(map_chunk_task!(
        INPUT_CHUNK_SIZE = 2,
        OUTPUT_CHUNK_SIZE = 2,
        |buf: impl Iterator<Item = u32>| -> impl IntoIterator<Item = u32> {
            buf.map(|x| x + 1);
            return 1;
        }
    ));
}
