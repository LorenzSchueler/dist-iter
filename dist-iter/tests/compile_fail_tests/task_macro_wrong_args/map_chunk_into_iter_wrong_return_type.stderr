error[E0271]: expected `{closure@map_chunk_into_iter_wrong_return_type.rs:7:85}` to be a closure that returns `u32`, but it returns `i32`
 --> tests/compile_fail_tests/task_macro_wrong_args/map_chunk_into_iter_wrong_return_type.rs:4:48
  |
4 |       [1, 2, 3, 4, 5].into_iter().dist_map_chunk(map_chunk_task!(
  |  ________________________________________________^
5 | |         INPUT_CHUNK_SIZE = 2,
6 | |         OUTPUT_CHUNK_SIZE = 2,
7 | |         |buf: impl Iterator<Item = i32>| -> impl IntoIterator<Item = u32> { buf.map(|x| x + 1) }
8 | |     ));
  | |_____^ expected `i32`, found `u32`
  |
  = note: required for `Map<impl Iterator<Item = i32>, {closure@map_chunk_into_iter_wrong_return_type.rs:7:85}>` to implement `Iterator`
  = note: consider using `--verbose` to print the full type name to the console
  = note: this error originates in the macro `::dist_iter::task` which comes from the expansion of the macro `map_chunk_task` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: `{integer}` is not an iterator
  --> tests/compile_fail_tests/task_macro_wrong_args/map_chunk_into_iter_wrong_return_type.rs:9:48
   |
9  |       [1, 2, 3, 4, 5].into_iter().dist_map_chunk(map_chunk_task!(
   |  ________________________________________________^
10 | |         INPUT_CHUNK_SIZE = 2,
11 | |         OUTPUT_CHUNK_SIZE = 2,
12 | |         |buf: impl Iterator<Item = u32>| -> impl IntoIterator<Item = u32> {
...  |
15 | |         }
16 | |     ));
   | |_____^ `{integer}` is not an iterator
   |
   = help: the trait `Iterator` is not implemented for `{integer}`, which is required by `{integer}: IntoIterator`
   = note: if you want to iterate between `start` until a value `end`, use the exclusive range syntax `start..end` or the inclusive range syntax `start..=end`
   = note: required for `{integer}` to implement `IntoIterator`
   = note: this error originates in the macro `::dist_iter::task` which comes from the expansion of the macro `map_chunk_task` (in Nightly builds, run with -Z macro-backtrace for more info)
